.travis.yml

```diff
@@ -1,26 +1,51 @@
 language: python
-python:
-- 3.6
-- 3.5
+matrix:
+  include:
+    - sudo: required
+      python: 3.5
+      services:
+        - docker
+      env:
+        - CIBW_BUILD=cp35-*
+        - PIP=pip
+        - PYTHON=python
+    - sudo: required
+      python: 3.6
+      services:
+        - docker
+      env:
+        - CIBW_BUILD=cp36-*
+        - PIP=pip
+        - PYTHON=python
 env:
   global:
-    - TRAVIS_RUST_VERSION=nightly
-    - RUST_BACKTRACE=1
-
+  - TRAVIS_RUST_VERSION=nightly
+  - RUST_BACKTRACE=1
+  - TWINE_USERNAME=pattonw
+  
 addons:
   apt:
     packages:
-      - libcurl4-openssl-dev
-      - libelf-dev
-      - libdw-dev
-      - cmake
-      - gcc
-      - binutils-dev
-      - libiberty-dev
-
-# add rust
+    - libcurl4-openssl-dev
+    - libelf-dev
+    - libdw-dev
+    - cmake
+    - gcc
+    - binutils-dev
+    - libiberty-dev
 before_install:
-  - source ./ci/travis/setup.sh
+- source ./ci/travis/setup.sh
+install:
+- pip install -U tox-travis cibuildwheel setuptools_rust
+script:
+- tox
+- |
+  if [[ $TRAVIS_TAG ]]; then
```

Comment:
I'd still think about putting this in an `after_success` block: failures in the `script` block don't stop the build, so the current setup allows you to deploy code which fails the tests.

I think you should also be able to put all the repeated stuff from the `matrix` into the global namespace (everything except `python:` and `CIBW_BUILD=`). Saves you some lines and makes it a bit more readable.

Strictly speaking, you could avoid the need for the matrix at all by setting `CIBW_BUILD` based on the python version (albeit with some slightly-gross string mangling):

```bash
export CIBW_BUILD=cp$(python -c "import sys; print(*sys.version_info[:2], sep='')")-*
```

or

```bash
export CIBW_BUILD=cp$(python -V | awk '{split($2,a,"."); print a[1] a[2]}')-*
```

Reply:
Yeah at the moment the .yml is not terribly well written. I ran into a couple problems that I didn't want to handle right now. 
1) I'm not sure if you want an all or nothing approach, i.e. if python3.6 tests pass, should it build the python3.6 version even if the python3.5 tests failed?
2) I don't know how to split up tox tests by version, right now its running all the tests for every job. The designated python version in the .yml doesn't seem to do anything since the tox file designates both.
3) I think Tags are necessary for PyPi to have versions and be happy, but I'm also not really sure how tags work in git. If you tag a commit, and it fails the tests and doesn't build, is it a pain to fix the bugs and retag, or do you just bump the version again? If you're using bump2version, it seems annoying to redo a version bump.

It was in a minimally working mode, I figured I would leave it at that for now. I thought I could get around checking that tests pass in the build phase by only tagging aesthetic commits that change version number strings when everything else is working.

Reply:
> I'm not sure if you want an all or nothing approach, i.e. if python3.6 tests pass, should it build the python3.6 version even if the python3.5 tests failed?

That's a good point. I *think* that `after_success` happens only after all jobs complete: [this page](https://docs.travis-ci.com/user/job-lifecycle#the-job-lifecycle) states that `after_success` can access the result of the *build*, and that one build can be comprised of many jobs; suggesting that the build is the whole bunch which get kicked off together. That may mean it only happens once, I'm not sure. You can get the [deploy](https://docs.travis-ci.com/user/deployment/script/) stage to run an arbitrary script, but again, not sure how many times per build that can happen.

I think all-or-nothing is a good choice: everything should pass on all supported versions before a release gets deployed. 

> I don't know how to split up tox tests by version, right now its running all the tests for every job. The designated python version in the .yml doesn't seem to do anything since the tox file designates both.

It should do. `tox-travis`, when installed, effectively changes the `tox` command so that if there's a `TRAVIS_PYTHON_VERSION`, tox will only run the environment which has a python version matching that spec. I think this is working - looking at the build log, the 3.5 job is only running the py35 environment, and the 3.6 job is only running the py36 environment. If your tox is cycling through something else as well as the python version (like a variant of an environment variable), it'll run all matching environments ([catpy](https://travis-ci.org/catmaid/catpy) does this - it runs 5 jobs, one for each python version, and each one runs 2 tox environments).

> I think Tags are necessary for PyPi to have versions and be happy, but I'm also not really sure how tags work in git. If you tag a commit, and it fails the tests and doesn't build, is it a pain to fix the bugs and retag, or do you just bump the version again? If you're using bump2version, it seems annoying to redo a version bump.

Tags are just a reference to a commit (so are branches). The commit doesn't contain the tag, which is why pushing the commits doesn't push the tags by default. The simplest thing to do is what you're doing now: make all your changes, push them, make sure the build passes, and then do your version-bumping, and push that. To be really safe you can push the version-bump without the tags, make sure that changing the version string didn't break the build, and then push the tags, which will start the build again with exactly the same commit, but tagged this time. You can also move tags so it's possible (if awkward) to push the bump2version commit (but not the tag), then if the build fails, do your fixes, `rebase -i` to move the bump2version commit to the end, move the tag as well if necessary ('m not sure).

Or you can do what I do, which is forget how to push tags properly, and make ~3 patch versions every time I want to release anything.

> It was in a minimally working mode, I figured I would leave it at that for now.

Not a problem! It's great you got it working. I may raise an issue with these points just so they're in the backlog for future reference.